cupy/core/core.pyx

```diff
@@ -3060,162 +3085,67 @@ cpdef _scatter_op_mask_single(ndarray a, ndarray mask, v, Py_ssize_t axis, op):
-            if not isinstance(s, slice) or s != slice(None):
-                n_not_slice_none += 1
-                if issubclass(s.dtype.type, numpy.bool_):
-                    mask_i = i
-        if n_not_slice_none != 1:
-            raise ValueError('currently, CuPy only supports slices that '
-                             'consist of one boolean array.')
-        _scatter_op_mask_single(a, slices[mask_i], value, mask_i, op)
+        mask_i = _get_mask_index(slice_list)
+        _scatter_op_mask_single(a, slice_list[mask_i], value, mask_i, op)
         return
 
     if advanced:
-        # split slices that can be handled by basic-indexing
-        basic_slices = []
-        adv_slices = []
-        for i, s in enumerate(slices):
-            if type(s) is slice:
-                basic_slices.append(s)
-                adv_slices.append(slice(None))
-            elif s is None:
-                basic_slices.append(None)
-                adv_slices.append(slice(None))
-            elif (isinstance(s, ndarray) and
-                    issubclass(s.dtype.type, numpy.integer)):
-                basic_slices.append(slice(None))
-                adv_slices.append(s)
-            elif isinstance(s, int):
-                basic_slices.append(slice(None))
-                scalar_array = ndarray((), dtype=numpy.int64)
-                scalar_array.fill(s)
-                adv_slices.append(scalar_array)
-            else:
-                raise IndexError(
-                    'only integers, slices (`:`), ellipsis (`...`),'
-                    'numpy.newaxis (`None`) and integer or '
-                    'boolean arrays are valid indices')
-
-        # check if this is a combination of basic and advanced indexing
-        for s in basic_slices:
-            if s is None or (isinstance(s, slice) and s != slice(None)):
-                # returns a view of a
-                a = a[tuple(basic_slices)]
-                break
-
-        arr_slices_mask = [not isinstance(s, slice) for s in adv_slices]
-        if sum(arr_slices_mask) == 1:
-            axis = arr_slices_mask.index(True)
-            _scatter_op_single(a, adv_slices[axis], value,
-                               li=axis, ri=axis, op=op)
+        a, adv_slices, adv_mask = _prepare_advanced_indexing(a, slice_list)
+        if sum(adv_mask) == 1:
+            axis = adv_mask.index(True)
+            _scatter_op_single(a, adv_slices[axis], value, axis, axis, op)
             return
 
         # scatter_op with multiple integer arrays
         a_interm, reduced_idx, li, ri =\
             _prepare_multiple_array_indexing(a, adv_slices)
-        _scatter_op_single(a_interm, reduced_idx, value, li=li, ri=ri, op=op)
+        _scatter_op_single(a_interm, reduced_idx, value, li, ri, op)
         return
 
+    y = _simple_getitem(a, slice_list)
     if op == 'update':
-        v = a[tuple(slices)]
-        if isinstance(value, ndarray):
-            y, x = broadcast(v, value).values
```

Comment:
This broadcast is removed in the PR. Is it OK?

Reply:
Yes, the array is broadcated by elementwise kernel.